event/src/lib.rs

```diff
@@ -1,32 +1,56 @@
-//! which utilizes `crossbeam-channel` to support "blocking until an event arrives".
-//!
-//! In this documentation, all items marked with `channels` are only available if the
-//! `crossbeam-channel` feature is enabled.
-//!
-//! ## Cascades
-//!
-//! Sometimes, it is necessary to route events between multiple threads and event queues.
-//! When the `crossbeam-channel` feature is enabled, this crate offers the `cascade` API,
-//! which supports filtered event forwarding.
+/*!
+
+```text
+╭─── RawEventQueue
+│
+│ ╔═════════════════╦════════════════════╦══════════════════╗
+│ ║  sending halves ║     forwarders     ║ receiving halves ║
+│ ╟─────────────────╫────────────────────╫──────────────────╢
+│ ║ ╭───────────────╫────────────────────╫────────────╮     ║
+│ ║ │               ║                    ║  (push_*)  ↑     ║
+│ ║ │               ║ ╭───── cascade::* ─╫──┬filter───╯     ║
+↕ ║ ↓               ║ ↑                  ║  ╰*...─────╯     ║
+│ ║ │               ║ │                  ║                  ║
+╰─╫─┼─ Emitter[Mut]─╫─┴─ if Listable ─>>─╫──┬─ Listen       ║
+  ╠═╪═══════════════╩════════════════════╬══╪═════════════╤═╩═══════════════╗
+  ║ ├─ nested Emitters...                ║  ├─ merge::*   │  multiplexing/  ║
+  ║ │   (broadcasting)                   ║  │  (muxing)   │  broadcasting   ║
+  ╚═╪════════════════════════════════════╩══╪═════════════╧═════════════════╝
+    ↑                                       ↓
+    ╰─ emit(_Event)                         ╰─ with(_f(&[_Event]))
+```
+
+# `Emitter`'s
+
+`Emitter`'s (data types which implement the [`EventEmitter`](prelude::EventEmitter) trait)
+represent an event queue instance including the event sending capability
+and often (when they implement the [`QueueInterfaceListable`](traits::QueueInterfaceListable))
+have the ability to construct new `Listener`s.
+
+# `Listener`'s
+
+`Listener`'s (data types which implement the [`EventListen`](prelude::EventListen) trait)
+represent the receiving end of an event queue.
+A `Listener` should be wrapped inside of an Rc or Arc
+if you want multiple references to the same listener
+
+# `crossbeam-channel` support
+
+This crate offers a simple non-blocking API by default. But this isn't enough in multi-threaded
+scenarios, because often polling/busy-waiting inside of non-main-threads isn't wanted and wastes
+resources. Thus, this crate offers an blocking API if the `crossbeam-channel` feature is enabled,
+which utilizes `crossbeam-channel` to support "blocking until an event arrives".
+
+In this documentation, all items marked with `channels` are only available if the
+`crossbeam-channel` feature is enabled.
+
+## Cascades
+
+Sometimes, it is necessary to route events between multiple threads and event queues.
+When the `crossbeam-channel` feature is enabled, this crate offers the `cascade` API,
+which supports filtered event forwarding.
+
+*/
 
 #![cfg_attr(feature = "docs", feature(doc_cfg))]
```

Comment:
Side note: this line causes an error when trying to build with `--all-features`.

Reply:
This feature is designed to only be enabled when documentation is generated, the feature is provided by nightly rustdoc (at least).

Reply:
related issue: https://github.com/tokio-rs/tracing/issues/440